canonical/main.go

```diff
@@ -0,0 +1,377 @@
+	pkg := parsePkg(name)
+
+	obj := inBucket.Object(name)
+
+	// Storage reader -> un-gzip -> un-tar
+	// -> process
+	// -> tar -> gzip -> Storage writer.
+	r, err := obj.NewReader(ctx)
+	if err != nil {
+		return fmt.Errorf("unable to read from gs://%s/%s: %v", obj.BucketName(), name, err)
+	}
+
+	gr, err := gzip.NewReader(r)
+	if err != nil {
+		log.Fatal(err)
+	}
+	tr := tar.NewReader(gr)
+
+	outObj := outBucket.Object(name)
+	ow := outObj.NewWriter(ctx)
+	gw := gzip.NewWriter(ow)
+	tw := tar.NewWriter(gw)
+
+	// Loop over every header in the tarball.
+	for {
+		hdr, err := tr.Next()
+		if err == io.EOF {
+			break // End of archive
+		}
+
+		if err != nil {
+			return fmt.Errorf("error reading block from gs://%s/%s: %v", obj.BucketName(), name, err)
+		}
+
+		// Don't need to do anything with directory headers.
+		if hdr.FileInfo().IsDir() {
+			tw.WriteHeader(hdr)
+			continue
+		}
+
+		// Only process .html files.
+		if !strings.HasSuffix(hdr.Name, ".html") {
+			tw.WriteHeader(hdr)
+			if l, err := io.Copy(tw, tr); err != nil || l != hdr.Size {
+				return fmt.Errorf("error writing %q %q: wrote %v bytes: %v", name, hdr.Name, l, err)
+			}
+			continue
+		}
+
+		// Scanner to read every line.
+		scanner := bufio.NewScanner(tr)
+		// Write to a buffer so we can get the new number of bytes for the record.
+		out := &bytes.Buffer{}
+
+		// foundCanonical and inHead keep track of where we are, naively,
+		// in the HTML file.
+		foundCanonical := false
+		inHead := false
+
+		filename := hdr.Name[len("./"):]
+
+		// Note: this does not preserve line endings.
+		// You can `diff --ignore-space-change` to verify there
+		// are no changes to content.
+		for scanner.Scan() {
+			// Read as bytes to avoid an unnecessary string allocation.
+			text := scanner.Bytes()
+
+			// Trim space rather than use regex.
+			trimmed := bytes.TrimSpace(text)
+			if bytes.Equal(trimmed, startHead) {
+				inHead = true
+			}
+
+			if inHead && bytes.Contains(text, canonical) {
+				foundCanonical = true
+			}
+
+			if inHead && bytes.Equal(trimmed, endHead) {
+				if !foundCanonical {
+					link, err := canonicalLink(ctx, pkg, filename)
+					if err != nil {
+						return fmt.Errorf("canonicalLink: %v", err)
+					}
+					fmt.Fprintf(out, "    <link rel=\"canonical\" href=%q/>\n", link)
+					foundCanonical = true
+				}
+				inHead = false
+			}
+
+			out.Write(text)
+			fmt.Fprintln(out) // Text does not have a newline. Add one.
+		}
+		if err := scanner.Err(); err != nil {
+			return fmt.Errorf("error scanning %q %q: %v", name, hdr.Name, err)
+		}
+
+		// Update the header to have the right size. Otherwise, leave hdr unchanged.
+		hdr.Size = int64(out.Len())
+		tw.WriteHeader(hdr)
+		if l, err := tw.Write(out.Bytes()); err != nil || l != int(hdr.Size) {
+			return fmt.Errorf("error writing %q %q: wrote %v bytes: %v", name, hdr.Name, l, err)
+		}
+
+		atomic.AddInt64(&numFiles, 1)
+		bar.Set("prefix", fmt.Sprintf("Canonicalizing... %d files | ", numFiles))
```

Comment:
Does numFiles need to be loaded, honest question?

Reply:
I'm not sure what you mean by loaded. But, it's very nice to see the number of files go up as the tool is running. Otherwise, it's not clear if processing is stalled.

In one case, I had a deadlock in the code, didn't have this progress prefix, and didn't know it was stuck until I ctrl+c'd.

Reply:
Sorry I should have been more specific I meant [atomic.LoadInt64](https://pkg.go.dev/sync/atomic#LoadInt64)